File: go/pass7.go
package main

import (
	"fmt"
)

// Check if two adjacent nodes are also adjacent in the other file
func (i *Ifcomp) pass7CombineAdjacentNodes(node1 TreeIndex) bool {
	firstIdx := toArrayIndex(First)
	secondIdx := toArrayIndex(Second)

	node2 := i.TreeState.Node[node1].Next

	if node2 == i.TreeState.Trees[firstIdx].End {
		return false
	}

	if i.DebugDumpTreesFull {
		i.printf("combine node1=%d ln=%d to node2=%d ln=%d\n",
			node1, i.TreeState.Node[node1].Linen, node2, i.TreeState.Node[node2].Linen)
	}

	findIdx := TreeIndex(i.FileState.FileLine[firstIdx][i.trueLineOf(node1)].Ptr0)
	findNodeI := i.findNode(i.TreeState.Trees[secondIdx], findIdx)

	findIdx2 := TreeIndex(i.FileState.FileLine[firstIdx][i.trueLineOf(node2)].Ptr0)
	findNodeJ := i.findNode(i.TreeState.Trees[secondIdx], findIdx2)

	if findNodeI == NullNode || findNodeJ == NullNode {
		return false
	}

	if findNodeJ == i.TreeState.Node[findNodeI].Next {
		i.combineNodes(node1, node2)
		i.combineNodes(findNodeI, findNodeJ)
		return true
	}
	return false
}

// Combine two adjacent nodes into a branch structure
func (i *Ifcomp) combineNodes(node1, node2 TreeIndex) {
	var branchLink1, branchLink2 TreeIndex
	N := NodeDecl{}
	N.Cost = i.TreeState.Node[node1].Cost + i.TreeState.Node[node2].Cost
	N.Linen = i.TreeState.Node[node1].Linen

	// First remove node2 from file2
	i.detachNode(node2)
	N.Prev = i.TreeState.Node[node1].Prev
	N.Next = i.TreeState.Node[node1].Next

	// Now remove node1 from file1
	i.detachNode(node1)

	if !i.leaf(node1) {
		// Just want the branch
		N.BranchStart = i.TreeState.Node[node1].BranchStart
		branchLink1 = i.TreeState.Node[node1].BranchEnd
		// The sequence in node1 is absorbed in N and hence isn't needed
		i.freeNode(node1)
		node1 = N.BranchStart
	} else {
		N.BranchStart = node1
		branchLink1 = node1
	}

	if !i.leaf(node2) {
		branchLink2 = i.TreeState.Node[node2].BranchStart
		N.BranchEnd = i.TreeState.Node[node2].BranchEnd
		// The sequence in node2 is absorbed in N and hence isn't needed
		i.freeNode(node2)
		node2 = branchLink2
	} else {
		branchLink2 = node2
		N.BranchEnd = node2
	}

	newNode := i.makeNode(N)
	// Insert new_node after N.prev and before N.next; i.e., it replaces node1
	i.TreeState.Node[N.Prev].Next = newNode
	i.TreeState.Node[N.Next].Prev = newNode
	i.TreeState.Node[N.BranchStart].Prev = newNode
	i.TreeState.Node[N.BranchEnd].Next = newNode
	i.TreeState.Node[branchLink1].Next = branchLink2
	i.TreeState.Node[branchLink2].Prev = branchLink1
}

// Free a node
func (i *Ifcomp) freeNode(n TreeIndex) {
	if i.DebugDontFree {
		return
	}
	i.TreeState.Node[n].Next = i.TreeState.FreeNodesStart
	i.TreeState.FreeNodesStart = n
}

// Pass 7: Combine Adjacent Nodes
//
// Purpose: Merge adjacent nodes that are also adjacent in the other file,
// reducing tree complexity and creating larger matched segments.
func (i *Ifcomp) pass7() error {
	firstIdx := toArrayIndex(First)
	nodeIdx := i.TreeState.Node[i.TreeState.Trees[firstIdx].Start].Next
	endIdx := i.TreeState.Trees[firstIdx].End

	// Safety check: prevent infinite loops
	iterationCount := 0
	const MAX_ITERATIONS = 10000

	// Loop until we reach the end marker
	// Check both: nodeIdx itself and node[nodeIdx].next
	for nodeIdx != endIdx && i.TreeState.Node[nodeIdx].Next != endIdx {
		iterationCount++
		if iterationCount > MAX_ITERATIONS {
			return fmt.Errorf("internal error in pass7: infinite loop detected at node %d after %d iterations",
				nodeIdx, iterationCount)
		}

		// Safety check: if we've reached an invalid node (header or null), exit
		if nodeIdx == i.TreeState.Trees[firstIdx].Start || nodeIdx == NullNode || nodeIdx == 0 {
			break
		}

		j := i.TreeState.Node[nodeIdx].Prev
		if i.pass7CombineAdjacentNodes(nodeIdx) {
			// After combination, nodeIdx (node1) may have been freed/reused
			// Use j (prev) to get the new combined node
			nodeIdx = i.TreeState.Node[j].Next
		} else {
			// Advance to next node, but ensure we don't go to an invalid node
			nextNode := i.TreeState.Node[nodeIdx].Next
			if nextNode == NullNode || nextNode == 0 || nextNode == i.TreeState.Trees[firstIdx].Start {
				break
			}
			nodeIdx = nextNode
		}
	}

	return nil
}

File: rust/src/pass7.rs
use crate::types::*;
use super::Ifcomp;

impl Ifcomp {
    // Check if two adjacent nodes are also adjacent in the other file
    pub fn pass7_combine_adjacent_nodes(&mut self, node1: TreeIndex) -> bool {
        let first_idx = FileIndex::First.to_array_index();
        let second_idx = FileIndex::Second.to_array_index();

        let node2 = self.tree_state.node[node1].next;

        if node2 == self.tree_state.trees[first_idx].end {
            return false;
        }

        if self.debug_dump_trees_full {
            println!("combine node1={} ln={} to node2={} ln={}\n",
                node1, self.tree_state.node[node1].linen, node2, self.tree_state.node[node2].linen);
        }

        let find_idx = self.file_state.file_line[first_idx][self.true_line_of(node1) as usize].ptr0 as i16;
        let find_node_i = self.find_node(self.tree_state.trees[second_idx], find_idx);

        let find_idx2 = self.file_state.file_line[first_idx][self.true_line_of(node2) as usize].ptr0 as i16;
        let find_node_j = self.find_node(self.tree_state.trees[second_idx], find_idx2);

        if find_node_i == NULL_NODE || find_node_j == NULL_NODE {
            return false;
        }

        if find_node_j == self.tree_state.node[find_node_i].next {
            self.combine_nodes(node1, node2);
            self.combine_nodes(find_node_i, find_node_j);
            return true;
        }
        false
    }

    // Pass 7: Combine Adjacent Nodes
    //
    // Purpose: Merge adjacent nodes that are also adjacent in the other file,
    // reducing tree complexity and creating larger matched segments.
    pub fn pass7(&mut self) -> Result<(), String> {
        let first_idx = FileIndex::First.to_array_index();
        let mut node_idx = self.tree_state.node[self.tree_state.trees[first_idx].start].next;
        let end_idx = self.tree_state.trees[first_idx].end;

        // Safety check: prevent infinite loops
        let mut iteration_count = 0;
        const MAX_ITERATIONS: usize = 10000;

        // Loop until we reach the end marker
        // Check both: node_idx itself and node[node_idx].next
        while node_idx != end_idx && self.tree_state.node[node_idx].next != end_idx {
            iteration_count += 1;
            if iteration_count > MAX_ITERATIONS {
                return Err(format!("internal error in pass7: infinite loop detected at node {} after {} iterations",
                    node_idx, iteration_count));
            }

            // Safety check: if we've reached an invalid node (header or null), exit
            if node_idx == self.tree_state.trees[first_idx].start || node_idx == NULL_NODE || node_idx == 0 {
                break;
            }

            let j = self.tree_state.node[node_idx].prev;
            if self.pass7_combine_adjacent_nodes(node_idx) {
                // After combination, node_idx (node1) may have been freed/reused
                // Use j (prev) to get the new combined node
                node_idx = self.tree_state.node[j].next;
            } else {
                // Advance to next node, but ensure we don't go to an invalid node
                let next_node = self.tree_state.node[node_idx].next;
                if next_node == NULL_NODE || next_node == 0 || next_node == self.tree_state.trees[first_idx].start {
                    break;
                }
                node_idx = next_node;
            }
        }

        Ok(())
    }
}
